pkg/asset/kubeconfig/kubeconfig_test.go

```diff
@@ -0,0 +1,229 @@
+`),
+		},
+		{
+			userName: "kubelet",
+			certKey:  kubeletCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:         rootCAState,
+				kubeletCertKey: kubeletCertState,
+				installConfig:  installConfigState,
+			},
+			errString: "",
+			expectedData: []byte(`clusters:
+- cluster:
+    certificate-authority-data: VEhJUyBJUyBST09UIENBIENFUlQgREFUQQ==
+    server: https://test-cluster-name-api.test.example.com:6443
+  name: test-cluster-name
+contexts:
+- context:
+    cluster: test-cluster-name
+    user: kubelet
+  name: kubelet
+current-context: kubelet
+preferences: {}
+users:
+- name: kubelet
+  user:
+    client-certificate-data: VEhJUyBJUyBLVUJFTEVUIENFUlQgREFUQQ==
+    client-key-data: VEhJUyBJUyBLVUJFTEVUIEtFWSBEQVRB
+`),
+		},
+		{
+			userName: "admin", // No root ca in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				adminCertKey:  adminCertState,
+				installConfig: installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "admin", // No admin cert in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:         rootCAState,
+				kubeletCertKey: kubeletCertState,
+				installConfig:  installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "kubelet", // No kubelet cert in parents.
+			certKey:  kubeletCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:        rootCAState,
+				adminCertKey:  adminCertState,
+				installConfig: installConfigState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+		{
+			userName: "admin", // No install config in parents.
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:       rootCAState,
+				adminCertKey: adminCertState,
+			},
+			errString:    "failed to find kubeconfig.fakeAsset in parents",
+			expectedData: nil,
+		},
+	}
+
+	for i, tt := range tests {
+		kubeconfig := &Kubeconfig{
+			rootDir:       testDir,
+			userName:      tt.userName,
+			rootCA:        rootCA,
+			certKey:       tt.certKey,
+			installConfig: installConfig,
+		}
+		st, err := kubeconfig.Generate(tt.parents)
+		if err != nil {
+			assert.EqualErrorf(t, err, tt.errString, fmt.Sprintf("test #%d expected %v, saw %v", i, tt.errString, err))
+			continue
+		} else {
+			if tt.errString != "" {
+				t.Errorf("test #%d expect error %v, saw nil", i, tt.errString)
+			}
+		}
+
+		filename := filepath.Join(testDir, "auth", fmt.Sprintf("kubeconfig-%s", tt.userName))
+		data, err := ioutil.ReadFile(filename)
+		if err != nil {
+			t.Errorf("test #%d failed to read kubeconfig %q: %v", i, filename, err)
+		}
+
+		if !bytes.Equal(st.Contents[0].Data, data) {
+			t.Errorf("test #%d expect kubeconfig data: %q, saw %q", i, string(st.Contents[0].Data), string(data))
```

Comment:
Using [`assert.Equal`][1] may give you diffs when these fail, although I haven't tested that.

[1]: https://godoc.org/github.com/stretchr/testify/assert#Equal